feat(ivy): add injectable-pipe schematic

crisbeto opened the pull request:
Adds a schematic that will annotate all `Pipe` classes with `Injectable` so that they can be injected.

This PR resolves FW-1228.

devversion commented on packages/core/schematics/migrations.json:
```diff
@@ -9,6 +9,11 @@
       "factory": "./migrations/template-var-assignment/index"
+    },
+    "migration-v8-injectable-pipe": {
+      "version": "8",
```
Should be `8-beta`.

devversion commented on packages/core/schematics/migrations/injectable-pipe/BUILD.bazel:
```diff
@@ -0,0 +1,19 @@
+        "//packages/core/schematics/test:__pkg__",
+    ],
+    deps = [
+        "//packages/compiler",
```
Is this migration actually using `@angular/compiler`?

devversion commented on packages/core/schematics/migrations/injectable-pipe/google3/injectablePipeRule.ts:
```diff
@@ -0,0 +1,61 @@
+          failures.push(new RuleFailure(
+              sourceFile, namedImports.getStart(), namedImports.getWidth(),
+              'Injectable needs to be imported in order to annotate Pipe classes.', this.ruleName,
+              importFix));
```
Thinking about this. Do we really want to create two failures for that? If we look at this from a linting perspective, I'd just expect a failure saying that there needs to be the `@Injectable` decorator. The user is responsible for adding the import then.

The replacement for the import can just be added to the `RuleFailure` for adding `@Injectable`. I don't think it would be nice to have two failures for a single issue (not sure how this affects Google3 and the CLs)

crisbeto replied:
I was a bit on the fence about this as well, but AFAIK we can't have the fixer for it without tying it to a failure.

devversion replied:
My thinking was to chain the both replacements into one failure. This is possible because `tslint.Fix` accepts an array of replacements.

crisbeto replied:
Ah, nice, I hadn't noticed that it supports an array of replacements.

devversion commented on packages/core/schematics/migrations/injectable-pipe/util.ts:
```diff
@@ -0,0 +1,47 @@
+export function addNamedImport(node: ts.ImportDeclaration, importName: string) {
+  const namedImports = getNamedImports(node);
+
+  if (namedImports && ts.isNamedImports(namedImports)) {
```
Is the check for `isNamedImports` necessary here? Looks like `getNamedImports` should already be typed properly.

crisbeto replied:
`getNamedImports` technically covers it, but the strict null checks still complain about it.

devversion commented on packages/core/schematics/utils/tslint/BUILD.bazel:
```diff
@@ -4,5 +4,8 @@ ts_library(
     tsconfig = "//packages/core/schematics:tsconfig.json",
-    visibility = ["//packages/core/schematics/migrations/template-var-assignment/google3:__pkg__"],
+    visibility = [
+        "//packages/core/schematics/migrations/injectable-pipe/google3:__pkg__",
```
Does the `injectable-pipe` actually use the TSLint google3 utilities? The visibility is limited to the packages that explicitly use the utils (in order to enforce less unused dependencies; and more dependency awareness)

crisbeto replied:
Looks like it does. The test run fails without it like this:

```
ERROR: /Users/kkostadinov/projects/angular/packages/core/schematics/migrations/injectable-pipe/google3/BUILD.bazel:3:1: in ts_library rule //packages/core/schematics/migrations/injectable-pipe/google3:google3: target '//packages/core/schematics/utils/tslint:tslint' is not visible from target '//packages/core/schematics/migrations/injectable-pipe/google3:google3'. Check the visibility declaration of the former target if you think the dependency is legitimate
```

devversion replied:
It's because the Google3 build package still depends on it (while it is unused). Please see the new comment.

devversion reviewed (approved):
LGTM. Just a few minor nits.

kara reviewed (approved):
LGTM, once comments from @devversion are resolved

alfaproject commented:
Just wondering, but can't the `Pipe` decorator also have the `Injectable` decorator logic if it really is mandatory for Ivy anyway then it would be a 2 in 1 decorator

kara commented:
@alfaproject We considered that, but combining the decorators that way would force all future pipes to generate extra code for injection defs even if they aren't designed to be injected. If we keep them separate, pipes where injection doesn't make sense can save the bytes. 

Also worth noting that this schematic is a first pass. If we have time, we want to limit the pipes we're decorating to just those that are provided or injected.

crisbeto committed:
feat(ivy): add injectable-pipe schematic

Adds a schematic that will annotate all `Pipe` classes with `Injectable` so that they can be injected.

This PR resolves FW-1228.

crisbeto commented:
The feedback has been addressed @devversion @kara.

kara commented:
[presubmit](http://test/OCL:243299924:BASE:243299961:1555093495803:c109f3f5)

devversion commented on packages/core/schematics/migrations/injectable-pipe/google3/BUILD.bazel:
```diff
@@ -0,0 +1,13 @@
+    visibility = ["//packages/core/schematics/test:__pkg__"],
+    deps = [
+        "//packages/core/schematics/migrations/injectable-pipe",
+        "//packages/core/schematics/utils/tslint",
```
@crisbeto It shows that message because the `tslint` utils (which are not added) are still added here. Please remove that here and from the visibility.

crisbeto replied:
Fixed.

devversion replied:
thanks!

kara commented:
@crisbeto Can you rebase this?

crisbeto commented:
Rebased.

crisbeto committed:
fixup! feat(ivy): add injectable-pipe schematic
